gupnp_1_6: support cross compilation

uninsane opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
    - [x] also built `pkgsCross.aarch64-multiplatform.gupnp_1_6` from x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
  - the output hash for x86_64 doesn't change, so this should be inconsequential to non-cross dependents.
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

NickCao commented on pkgs/development/libraries/gupnp/1.6.nix:
```diff
@@ -46,8 +47,8 @@ stdenv.mkDerivation rec {
   mesonFlags = [
-    "-Dgtk_doc=true"
-    "-Dintrospection=true"
+    "-Dgtk_doc=${lib.boolToString (stdenv.buildPlatform == stdenv.hostPlatform)}"
+    "-Dintrospection=${lib.boolToString (stdenv.buildPlatform == stdenv.hostPlatform)}"
```
gobject-introspection actually works for cross compilation now, just put it in nativeBuildInputs. And if gtk_doc involves running binary for host platform, there is mesonEmulatorHook.

uninsane committed:
gupnp_1_6: support cross compilation

uninsane committed:
gssdp_1_6: support cross compilation

uninsane committed:
gssdp: enable introspection on cross builds

uninsane committed:
gupnp: enable introspection on cross builds

uninsane commented:
oh, this is showing `rebuild-linux: 501-1000`: should i retarget this against the staging branch?

NickCao commented:
> oh, this is showing `rebuild-linux: 501-1000`: should i retarget this against the staging branch?

Yes, and I'd like to wait for the `fetchpatch` changes to land.

Artturin commented:
Closing in favor of https://github.com/NixOS/nixpkgs/pull/223142 which adds a withIntrospection to support cross os compilation

added you to the co-authors of `gssdp{_1_6}: enable introspection when cross`

Artturin commented on pkgs/development/libraries/gupnp/1.6.nix:
```diff
@@ -36,6 +37,8 @@ stdenv.mkDerivation rec {
     vala
     gi-docgen
+  ] ++ lib.optionals (!stdenv.buildPlatform.canExecute stdenv.hostPlatform) [
+    mesonEmulatorHook
```
i'm able to build this without mesonEmulatorHook

uninsane commented:
opened upstream PRs and fixed this changset to use `fetchpatch` against those instead of ad-hoc patching. rebased against staging. i think i addressed everything?

- <https://gitlab.gnome.org/GNOME/gssdp/-/merge_requests/11>
- <https://gitlab.gnome.org/GNOME/gupnp/-/merge_requests/32>

@Artturin i'm not sure if you intended to close this as part of your comment above (you commented _without_ closing). if you want that, just cherry-pick the `fetchpatch` stuff into your PR & close this one. but since that one's in draft makes more sense to merge this one IMO.
